azure-iot-device/azure/iot/device/iothub/transport/mqtt/mqtt_provider.py

```diff
@@ -33,97 +40,112 @@ def __init__(self, client_id, hostname, username, ca_cert=None):
-            logger.info("UNSUBACK received for %s", str(mid))
+            self._resolve_pending_callback(mid)
+
+        def on_unsubscribe(client, userdata, mid):
+            logger.info("UNSUBACK received for {}".format(mid))
             # TODO: how to do failure?
-            try:
-                self.on_mqtt_unsubscribed(mid)
-            except:  # noqa: E722 do not use bare 'except'
-                logger.error("Unexpected error calling on_mqtt_unsubscribed")
-                logger.error(traceback.format_exc())
-
-        self._mqtt_client.on_connect = on_connect_callback
-        self._mqtt_client.on_disconnect = on_disconnect_callback
-        self._mqtt_client.on_publish = on_publish_callback
-        self._mqtt_client.on_subscribe = on_subscribe_callback
-        self._mqtt_client.on_message = on_message_callback
-        self._mqtt_client.on_unsubscribe = on_unsubscribe_callback
+            self._resolve_pending_callback(mid)
+
+        def on_publish(client, userdata, mid):
+            logger.info("payload published for {}".format(mid))
+            # TODO: how to do failed publish
+            self._resolve_pending_callback(mid)
+
+        def on_message(client, userdata, mqtt_message):
+            logger.info("message received on {}".format(mqtt_message.topic))
+            # MUST do LBYL here to avoid confusion with errors thrown in calling callback
+            if self.on_mqtt_message_received:
+                try:
+                    # TODO: Why is this returning _topic instead of topic?
+                    self.on_mqtt_message_received(mqtt_message.topic, mqtt_message.payload)
+                except:  # noqa: E722 do not use bare 'except'
+                    logger.error("Unexpected error calling on_mqtt_message_received")
+                    logger.error(traceback.format_exc())
+            else:
+                logger.warning(
+                    "No event handler callback set for on_mqtt_message_received - DROPPING MESSAGE"
+                )
+
+        self._mqtt_client.on_connect = on_connect
+        self._mqtt_client.on_disconnect = on_disconnect
+        self._mqtt_client.on_subscribe = on_subscribe
+        self._mqtt_client.on_unsubscribe = on_unsubscribe
+        self._mqtt_client.on_publish = on_publish
+        self._mqtt_client.on_message = on_message
 
         logger.info("Created MQTT provider, assigned callbacks")
 
     def connect(self, password):
         """
-        This method connects the upper transport layer to the mqtt broker.
+        Connect to the MQTT broker, using hostname and username set at instantiation.
+
         This method should be called as an entry point before sending any telemetry.
+
+        :param str password: The password for connecting with the MQTT broker.
         """
         logger.info("connecting to mqtt broker")
 
-        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+        ssl_context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLSv1_2)
```

Comment:
>ssl_context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLSv1_2) [](start = 8, length = 59)

Are we going to follow this pattern of specifying parameter name when calling methods with 1 required positional param ?
The guidelines say to do this only with more than 2.

Reply:
I honestly don't know why I did this. You're right, this isn't the pattern.

---
In reply to: [278271073](https://github.com/Azure/azure-iot-sdk-python-preview/pull/82#discussion_r278271073) [](ancestors = 278271073)